parcels/scripts/utils.py

```diff
@@ -0,0 +1,120 @@
+import numpy as np
+import xarray as xr
+from os import path
+from netCDF4 import Dataset
+
+
+def compute_curvilinear_rotation_angles(mesh_filename, rotation_angles_filename, variables, dimensions):
+
+    if path.isfile(rotation_angles_filename) and path.getmtime(rotation_angles_filename) > path.getmtime(mesh_filename):
+        return
+
+    dataset = xr.open_dataset(mesh_filename)
+    lonU = np.squeeze(getattr(dataset, dimensions['U']['lon']).values)
+    latU = np.squeeze(getattr(dataset, dimensions['U']['lat']).values)
+    lonV = np.squeeze(getattr(dataset, dimensions['V']['lon']).values)
+    latV = np.squeeze(getattr(dataset, dimensions['V']['lat']).values)
+    lonF = np.squeeze(getattr(dataset, dimensions['F']['lon']).values)
+    latF = np.squeeze(getattr(dataset, dimensions['F']['lat']).values)
+    dataset.close()
+
+    rad = np.pi / 180.
+    rpi = np.pi
+
+    zxnpu = - 2. * np.cos(rad*lonU[1:, 1:]) * np.tan(rpi/4. - rad*latU[1:, 1:]/2.)
+    zynpu = - 2. * np.sin(rad*lonU[1:, 1:]) * np.tan(rpi/4. - rad*latU[1:, 1:]/2.)
+    znnpu = zxnpu*zxnpu + zynpu*zynpu
+
+    zxnpv = - 2. * np.cos(rad*lonV[1:, 1:]) * np.tan(rpi/4. - rad*latV[1:, 1:]/2.)
+    zynpv = - 2. * np.sin(rad*lonV[1:, 1:]) * np.tan(rpi/4. - rad*latV[1:, 1:]/2.)
+    znnpv = zxnpv*zxnpv + zynpv*zynpv
+
+    zxffu = 2. * np.cos(rad*lonF[1:, 1:]) * np.tan(rpi/4. - rad*latF[1:, 1:]/2.) \
+        - 2. * np.cos(rad*lonF[:-1, 1:]) * np.tan(rpi/4. - rad*latF[:-1, 1:]/2.)
+    zyffu = 2. * np.sin(rad*lonF[1:, 1:]) * np.tan(rpi/4. - rad*latF[1:, 1:]/2.) \
+        - 2. * np.sin(rad*lonF[:-1, 1:]) * np.tan(rpi/4. - rad*latF[:-1, 1:]/2.)
+    znffu = np.sqrt(znnpu * (zxffu*zxffu + zyffu*zyffu))
+    znffu = np.maximum(znffu, 1.e-14)
+
+    zxffv = 2. * np.cos(rad*lonF[1:, 1:]) * np.tan(rpi/4. - rad*latF[1:, 1:]/2.) \
+        - 2. * np.cos(rad*lonF[1:, :-1]) * np.tan(rpi/4. - rad*latF[1:, :-1]/2.)
+    zyffv = 2. * np.sin(rad*lonF[1:, 1:]) * np.tan(rpi/4. - rad*latF[1:, 1:]/2.) \
+        - 2. * np.sin(rad*lonF[1:, :-1]) * np.tan(rpi/4. - rad*latF[1:, :-1]/2.)
+    znffv = np.sqrt(znnpv * (zxffv*zxffv + zyffv*zyffv))
+    znffv = np.maximum(znffv, 1.e-14)
+
+    gsinu = (zxnpu*zyffu - zynpu*zxffu) / znffu
+    gcosu = (zxnpu*zxffu + zynpu*zyffu) / znffu
+    gsinv = (zxnpv*zxffv + zynpv*zyffv) / znffv
+    gcosv = -(zxnpv*zyffv - zynpv*zxffv) / znffv
+
+    # ** netCDF4 writing, since xArray is bugged **
+    lonU = lonU[1:, 1:]
+    latU = latU[1:, 1:]
+    lonV = lonV[1:, 1:]
+    latV = latV[1:, 1:]
```

Comment:
why strip the first indices here?

Reply:
Since to compute an angle at a node, we need the position of this node and the previous one. So it means that we lose the first line and row of a mesh.
It is important to keep in mind, and is stated in the doc of the function